### PR TITLE
Refactor DIM IO layer

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
### Changelist 

Refactored the DIM IO layer to comply with the code architecture.

1. Created shared devices: `dev_rgbLed.h/c`, `dev_switch.h/c`
2. Created shared time source: `hw_time.h/c`
3. Renamed DIM IO modules (watchdog, sevensegs, stackwatermark) to be dev modules.
4. Had to do more significant refactoring around the app and dev sevenseg modules to comply with new architecture.
5. Removed old and outdated code.

NOTE: Tests still need to be refactored.

### Testing Done

- [x] CI passes
- [x] Tested on car

### CODE REVIEW

File: firmware/thruna/DIM/test/Test_StateMachine.cpp
Comment (will-chaba) [inline firmware/thruna/DIM/test/Test_StateMachine.cpp:1]: should be safe to rename to `test_stateMachine.cpp`, Cmake searches for *.cpp

File: firmware/thruna/DIM/test/Test_StateMachine.cpp [hunk 1]
```diff
@@ -186,7 +183,14 @@ class DimStateMachineTest : public BaseStateMachineTest
...
         RESET_FAKE(turn_off_pdm_status_led);
-        RESET_FAKE(io_led_enable);
+<<<<<<< HEAD
+<<<<<<< HEAD
+=======
+        // RESET_FAKE(dev_led_enable);
+>>>>>>> 2b062df5 (refactor dim io layer)
+=======
+        RESET_FAKE(dev_led_enable);
+>>>>>>> f51d0cfe (undo mistakes)
```
Comment (will-chaba) [inline firmware/thruna/DIM/test/Test_StateMachine.cpp:193]: leftover from merge conflic

File: scripts/code_generation/fakegen/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+create_fake_library(
+    "fakegen_test_fakes"
+    "${TEST_HDRS_TO_FAKE}"
+)
+compile_gtest_executable(
+    "fakegen_test"
+    "${TEST_SRCS}"
+    "${TEST_INCLUDE_DIRS}"
+)
+target_link_libraries("fakegen_test" "fakegen_test_fakes")
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/CMakeLists.txt:62]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/value_returning_func_no_params.cpp.j2 [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+void fake_{{ name }}_reset() 
+{
+    memset(&{{ func_return_value }}, 0U, sizeof({{ return_type }}));
+    {{ func_call_count }} = 0;
+}
+
+uint32_t fake_{{ name }}_callCount(void)
+{
+    return {{ func_call_count }};
+}
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/value_returning_func_no_params.cpp.j2:38]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/value_returning_func_no_params.hpp.j2 [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Set the value to be returned when {{ name }} is invoked.
+ */
+void fake_{{ name }}_returns({{return_type}} return_value);
+
+/**
+ * Return the number of times {{ name }} has been invoked.
+ */
+uint32_t fake_{{ name }}_callCount(void);
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/value_returning_func_no_params.hpp.j2:18]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/value_returning_func_params.cpp.j2 [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    );
+
+    // Return call count from map.
+    if({{ call_count_map }}.count(params)) {
+        return {{ call_count_map }}[params];
+    }
+    else {
+        return 0;
+    }
+}
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/value_returning_func_params.cpp.j2:154]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/value_returning_func_params.hpp.j2 [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Return the number of times {{ name }} has been invoked with a specific set of arguments.
+ */
+uint32_t fake_{{ name }}_callCountForArgs(
+{%- for param in params %}
+    {{ param.type }} {{ param.name }}
+    {%- if not loop.last %}, {% endif %}
+{%- endfor %}
+);
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/value_returning_func_params.hpp.j2:34]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/void_func_no_params.cpp.j2 [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+void fake_{{ name }}_reset() 
+{
+    {{ func_call_count }} = 0;
+}
+
+uint32_t fake_{{ name }}_callCount(void)
+{
+    return {{ func_call_count }};
+}
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/void_func_no_params.cpp.j2:26]: nit: missing newline

File: scripts/code_generation/fakegen/src/templates/void_func_no_params.hpp.j2 [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * Reset the fake for {{ name }}.
+ */
+void fake_{{ name }}_reset(void);
+
+/**
+ * Return the number of times {{ name }} has been invoked.
+ */
+uint32_t fake_{{ name }}_callCount(void);
```
Comment (will-chaba) [inline scripts/code_generation/fakegen/src/templates/void_func_no_params.hpp.j2:13]: I think you get the point.

**Review by will-chaba — APPROVED** [review]

File: firmware/thruna/DIM/test/Test_StateMachine.cpp
Comment (gtaharaedmonds) [reply firmware/thruna/DIM/test/Test_StateMachine.cpp:1]: done

File: firmware/thruna/DIM/test/Test_StateMachine.cpp (hunk 1 above)
Comment (gtaharaedmonds) [reply firmware/thruna/DIM/test/Test_StateMachine.cpp]: fixed

**Pull request merged**